File: synthesizer/codegen.py
from __future__ import annotations
from .ast_nodes import ASTNode, Hole
from .grammar import Production
from typing import Dict, List


def render(node: ASTNode, grammar: Dict[str, List[Production]]) -> str:
    """Render an AST node to a Rust source string."""
    if isinstance(node, Hole):
        return f"???:{node.nt}"
    if isinstance(node, str):
        return node

    prod = _find_production(node.kind, grammar)
    if prod is None:
        raise ValueError(f"No production for kind={node.kind!r}")

    if not prod.children_spec:
        return prod.rust_template

    rendered_children = [render(c, grammar) for c in node.children]
    return prod.rust_template.format(*rendered_children)


def wrap_in_harness(fn_src: str, fn_name: str, test_cases: List[Dict]) -> str:
    """
    Wrap function source in a main() that runs all test cases in one binary.
    Each call prints its result on a separate line.
    """
    calls = "\n".join(
        f'    println!("{{:?}}", {fn_name}({", ".join(tc["inputs"])}));'
        for tc in test_cases
    )
    return f"""{fn_src}

fn main() {{
{calls}
}}
"""


def _find_production(kind: str, grammar):
    for prods in grammar.values():
        for p in prods:
            if p.name == kind:
                return p
    return None

File: rust_synthesizer/src/codegen.rs
use crate::ast::{Node, Child};
use crate::grammar::{Grammar, find_production};

pub fn render(node: &Node, grammar: &Grammar) -> Result<String, String> {
    let prod = find_production(&node.kind, grammar)
        .ok_or_else(|| format!("No production for kind={}", node.kind))?;
    if prod.children_spec.is_empty() {
        return Ok(prod.rust_template.clone());
    }
    let parts: Result<Vec<String>, String> = node.children.iter().map(|c| match c {
        Child::Node(n) => render(n, grammar),
        Child::Hole(nt) => Ok(format!("???:{}", nt)),
    }).collect();
    Ok(format_template(&prod.rust_template, &parts?))
}

fn format_template(tmpl: &str, args: &[String]) -> String {
    let mut out = tmpl.to_string();
    // Replace in reverse index order to avoid {1} matching inside {10}
    for i in (0..args.len()).rev() {
        out = out.replace(&format!("{{{}}}", i), &args[i]);
    }
    out
}
